polynote-kernel/src/main/scala/polynote/kernel/remote/transport.scala

```diff
@@ -350,10 +365,14 @@ object SocketTransport {
     */
   // TODO: Maybe fs2.io.tcp.Socket methods could be made to work, just seems over-complicated for single-client server?
   // TODO: If this introduces allocation/GC latency, could try to use a shared, reused buffer
-  class FramedSocket(socketChannel: SocketChannel) {
+  class FramedSocket(socketChannel: SocketChannel, closed: Promise[Throwable, Unit]) {
     private val incomingLengthBuffer = ByteBuffer.allocate(4)
     private val outgoingLengthBuffer = ByteBuffer.allocate(4)
 
+    // using primitive j.u.concurrent Semaphore here, because I need tryAcquire (zio Semaphore doesn't have it)
+    // TODO: When zio Sempaphore has tryAcquire, use that instead
```

Comment:
All the changes here are basically so that we can have the keepalive be skipped if the channel is busy. Presumably ZIO will add a `Semaphore#tryAwait` at some point and it can clean this up a bit.